File: node/monitoring/monitor_client.js
const zmq = require('zeromq');

class MonitorClient {
    constructor(config) {
        this.host = config.monitoring.host;
        this.port = config.monitoring.metrics_pull_port;
        this.interval = config.monitoring.metrics_push_interval_sec * 1000;
        this.socket = new zmq.Push();
        this._connected = false;
        this._sending = false;
        this._queue = [];
    }

    async connect() {
        const address = `${this.host}:${this.port}`;
        this.socket.connect(address);
        this._connected = true;
        console.log(`[MonitorClient] Connected to ${address}`);
    }

    async sendMetrics(source, metrics) {
        if (!this._connected) return;

        const message = JSON.stringify({ source, metrics, timestamp: Date.now() });
        this._queue.push(message);
        this._flushQueue();
    }

    async _flushQueue() {
        if (this._sending || this._queue.length === 0) return;
        
        this._sending = true;
        while (this._queue.length > 0) {
            const message = this._queue.shift();
            try {
                await this.socket.send(message);
            } catch (err) {
                console.warn(`[MonitorClient] Push failed: ${err.message}`);
                // Put message back in queue
                this._queue.unshift(message);
                break;
            }
        }
        this._sending = false;
    }

    close() {
        if (this._connected) {
            this.socket.close();
            this._connected = false;
        }
    }
}

module.exports = { MonitorClient };

File: python/monitoring/monitor_client.py
import logging
import json
import zmq

logger = logging.getLogger(__name__)


class MonitorClient:
    """Wysyła metryki do Monitoring Service przez ZeroMQ PUSH."""

    def __init__(self, config):
        self.host = config['monitoring']['host']
        self.port = config['monitoring']['metrics_pull_port']
        self.interval = config['monitoring']['metrics_push_interval_sec']

        self.context = zmq.Context()
        self.socket = self.context.socket(zmq.PUSH)
        self.socket.setsockopt(zmq.SNDTIMEO, 2000)
        self.socket.setsockopt(zmq.LINGER, 0)

        self._connected = False

    def connect(self):
        address = f"{self.host}:{self.port}"
        self.socket.connect(address)
        self._connected = True
        logger.info(f"Monitor client connected to {address}")

    def send_metrics(self, source, metrics):
        if not self._connected:
            return

        try:
            message = json.dumps({'source': source, 'metrics': metrics})
            self.socket.send_string(message, zmq.NOBLOCK)
        except zmq.Again:
            logger.warning("Monitor push dropped (buffer full)")
        except Exception as e:
            logger.error(f"Monitor push error: {e}")

    def close(self):
        self.socket.close()
        self.context.term()
        self._connected = False
        logger.info("Monitor client closed.")
